src/fhirtypes/ElementDefinition.ts

```diff
@@ -417,7 +417,9 @@ export class ElementDefinition {
       .split('.')
       .map(p => {
         const i = p.indexOf('[x]:');
-        return i > -1 ? p.slice(i + 4) : p;
+        const isChoiceSlice =
+          i > -1 ? this.type?.some(type => p.slice(i + 4).endsWith(upperFirst(type.code))) : false;
```

Comment:
Since this iterates through the `id` split on `.`, I don't think this would work for deeper paths.  For example, something like `Observation.value[x]:valueQuantity.system` wouldn't work right when `p` is `value[x]:valueQuantity` because `type.code` would be `uri` (since `this.type` refers to the type of the element related to the _full_ id).

BUT... I'm a little worried about having to lookup the element for the intermediate id in this logic and check its type because that could have some effect on performance if its not done carefully.  So if that's the way you need to do it, please try to ensure that lookup only happens when absolutely needed.

For example, if it is `value[x]:fooBar`, we don't need to do a lookup because the slicename does not start with `value`, so it can't be the shortcut.